File: spine/orreth_spine/sinks.py
# PROVENANCE: Claude Fable 5 (claude-fable-5) — rearch P1 sp2, the events shadow (M4-lite) · 2026-09-16
"""The relay's real sinks (canon 0002).

KafkaSink publishes an outbox row to the events rail: the TOPIC is the
envelope's type (topics are schema families, never per-identity), the
KEY is the aggregate id when the envelope wears one — so order holds
exactly where it matters, per aggregate, and nowhere it doesn't. A
refused publish raises, the relay records the attempt, and the row
stays unpublished — at-least-once, honestly.
"""
from __future__ import annotations

import os

from confluent_kafka import Producer

from . import envelope as ev

KAFKA_BOOTSTRAP = os.environ.get("SPINE_KAFKA", "localhost:9092")


class KafkaSink:
    def __init__(self, bootstrap: str | None = None):
        self._p = Producer({"bootstrap.servers": bootstrap or KAFKA_BOOTSTRAP})

    def publish(self, message_id: str, body: bytes) -> None:
        env = ev.decode(body)
        topic = env["type"]
        key = str((env.get("aggregate") or {}).get("id") or message_id)
        errors: list = []
        self._p.produce(topic, value=body, key=key,
                        on_delivery=lambda e, _m: errors.append(e) if e else None)
        remaining = self._p.flush(10)
        if remaining or errors:
            raise ConnectionError(
                f"the events rail refused the publish: "
                f"{errors[0] if errors else 'flush timed out'}")

File: backend/plane/crates/orreth-spine/src/sinks.rs
// PROVENANCE: Claude Fable 5.1 (claude-fable-5-1) — rearch P7 sp2, the ground and the rails · 2026-09-22
//! The relay's real sink — mirrors `orreth_spine.sinks.KafkaSink` (canon
//! 0002). The TOPIC is the envelope's type (schema families, never
//! per-identity); the KEY is the aggregate id when the envelope wears one —
//! order holds exactly where it matters, per aggregate, and nowhere it
//! doesn't. A refused publish raises, the relay records the attempt, and the
//! row stays unpublished — at-least-once, honestly.

use crate::envelope;
use crate::outbox::Sink;
use crate::rail_error::RailError;
use crate::rails::{key_for, topic_for};
use rdkafka::config::ClientConfig;
use rdkafka::producer::{FutureProducer, FutureRecord};
use std::time::Duration;

pub struct KafkaSink {
    producer: FutureProducer,
    timeout: Duration,
}

impl KafkaSink {
    pub fn new(bootstrap: &str) -> Result<KafkaSink, RailError> {
        let producer: FutureProducer = ClientConfig::new()
            .set("bootstrap.servers", bootstrap)
            .set("message.timeout.ms", "10000")
            .create()?;
        Ok(KafkaSink {
            producer,
            timeout: Duration::from_secs(10),
        })
    }
}

impl Sink for KafkaSink {
    async fn publish(&mut self, message_id: &str, body: &[u8]) -> Result<(), RailError> {
        let env = envelope::decode(body)?;
        let topic = topic_for(&env)
            .ok_or_else(|| RailError::Refused("the envelope names no type".into()))?
            .to_string();
        let key = key_for(&env).unwrap_or_else(|| message_id.to_string());
        self.producer
            .send(
                FutureRecord::to(&topic).key(&key).payload(body),
                self.timeout,
            )
            .await
            .map(|_| ())
            .map_err(|(e, _)| RailError::Sink(format!("the events rail refused the publish: {e}")))
    }
}
